### PR TITLE
Instructor: edit 'Distribute points' questions: make status message span both label and input #7899

### LINKED ISSUE #7899: Instructor: edit 'Distribute points' questions: make status message span both label and input
In the `student feedback edit page`  under the `Distribute points (among recipients)` question, currently, the status message spans only the _input field_.

**Current Behaviour:**
Sample Image

**Expected:**
status message should span both _input label_ and _input field_.

### PR DESCRIPTION
**Pull request by joanneong**
Fixes #7899 partially

[Note: The issue has been extended to include rank questions as well, but this PR currently only deals with `constSum` questions. However, since their implementations will be similar, this PR has been opened first to clarify some doubts.]

**Outline of Solution**
The general solution which I have implemented is:
1. Treat `constSum` questions differently in `response.jsp`
    - Prevent the status message from being in the same `div` as the form input by removing the `div` element
2. Edit the template for `constSum` questions by modifying `ConstSumSubmissionFormTemplate` and `ConstSumSubmissionFormOptionFragment`
    - Limit the width of the `constSumMessage`
    - Limit the label length so that long questions do not affect input box (see below)
    - Centralise the text displayed by `constSumMessage` so that shorter text displays (e.g. 2 points left to distribute) will still look like it spans both the label and the input

However, I have a few queries that I hope reviewer(s) can help me with:

(a) The modification in `response.jsp`
Personally, I find my implementation a little too 'clunky' for my liking. I tried to use the equivalent of an `if-else` statement using `jstl`'s `c:choose` and `c:otherwise` tags, but this seems to be rather verbose. I am open to any better suggestions to implement this.

(b) Differing results for different `constSum` questions
There are actually two kinds of `constSum` questions that share the same display template: (i) Distribute points among recipients and (ii) Distribute points among options. Although the solution is the same for both kinds, the results actually differ a little. For all the images below, the left image is the original implementation, and the right image is my new implementation.

For **(i) Distribute points among recipients**, we have the following results:

Without points distribution:
combinedwithoutdist

---

With points distribution:
combinedwithdist

Meanwhile, for **(ii) Distribute points among options**, we have the following results:

Without points distribution:
combinedwithoutdist

---

With points distribution:
combinedwithdist

---

With really long options:
combinedlongoption

I find that the results for (i) Distribute points among recipients look better than the original in all the different scenarios. Whereas for (ii) Distribute points among options, I think that the original generally looks better, but the original comes with two "flaws" that my new implementation fixes. The first is that the original can result in (weird, to me) width changes in the display even in the same question (e.g. when the status message becomes longer, the input box automatically jumps and shifts further away from the input), but my implementation fixes this width. The second is that in the event of long options, the original will shift the entire option down to a new line, resulting in an inconsistent display whereas mine fixes the display (see images above). What do the reviewer(s) make of this?

### CODE REVIEW

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag [hunk 1]
```diff
@@ -49,12 +49,24 @@
...
-    <c:if test="${response.existingResponse}">
-      <input type="hidden"
-          name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_ID %>-${questionWithResponses.question.qnIndx}-${response.responseIndx}"
-          value="<c:out value="${response.responseId}"/>">
-    </c:if>
-  </div>
+  <c:choose>
+    <c:when test="${questionWithResponses.question.questionTypeConstsum}">
+      ${response.submissionFormHtml}
+      <c:if test="${response.existingResponse}">
```
Comment (xpdavid) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag]: The common fragment can be extracted out.

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag [hunk 2]
```diff
@@ -49,12 +49,24 @@
...
-  <div class="${divClassType}<c:if test="${questionWithResponses.question.questionTypeConstsum}"> width-auto</c:if>">
-    ${response.submissionFormHtml}
-    <c:if test="${response.existingResponse}">
-      <input type="hidden"
-          name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_ID %>-${questionWithResponses.question.qnIndx}-${response.responseIndx}"
-          value="<c:out value="${response.responseId}"/>">
-    </c:if>
-  </div>
+  <c:choose>
+    <c:when test="${questionWithResponses.question.questionTypeConstsum}">
```
Comment (xpdavid) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag]: To me, I can accept the solution :)

Maybe you could improve the arrangement to something like this?
image

To me, the width of the shouldn't be more than the content.
image

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag (hunk 1 above)
Comment (joanneong) [reply src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag]: Hello, I am not sure how to do extraction in a jsp file. I googled for a while, and realised that many websites suggest creating a jsp fragment/another tag file to do so, which may be an overkill for this instance. What do you think? Any better ideas? If not I will go ahead and create a jsp fragment...

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag (hunk 2 above)
Comment (joanneong) [reply src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/response.tag]: I have edited the files accordingly. Should I proceed to make similar changes for rank options questions as per the discussion in #7899 ?

File: src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
-  <div id="constSumInfo-${questionIndex}-${responseIndex}" class="form-group" ${constSumOptionVisibility}>
-    <div class="col-sm-12">
-      <hr class="margin-top-0">
-      <p class="text-color-blue" id="constSumMessage-${questionIndex}-${responseIndex}">
-      </p>
-      <hr>
-    </div>
+</div>
+<div id="constSumInfo-${questionIndex}-${responseIndex}" class="form-group col-sm-9" ${constSumOptionVisibility}>
+  <div class="col-sm-12 text-center">
```
Comment (xpdavid) [inline src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html]: Should be `col-sm-6` right? Also, the `col-sm-9` for parent `div` doesn't add much value so you can remove it.

This is what `col-sm-6` give you.

image

image

File: src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
   ${constSumSubmissionFormOptionFragments}
-  <div id="constSumInfo-${questionIndex}-${responseIndex}" class="form-group" ${constSumOptionVisibility}>
-    <div class="col-sm-12">
-      <hr class="margin-top-0">
-      <p class="text-color-blue" id="constSumMessage-${questionIndex}-${responseIndex}">
-      </p>
-      <hr>
-    </div>
+</div>
+<div id="constSumInfo-${questionIndex}-${responseIndex}" class="form-group col-sm-9" ${constSumOptionVisibility}>
```
Comment (xpdavid) [inline src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html]: Refer to the above comment, this is where you should add `clear:both` style.

File: src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html [hunk 5]
```diff
@@ -1,16 +1,16 @@
-<div>
+<div class="col-sm-6 auto-width">
```
Comment (xpdavid) [inline src/main/resources/feedbackQuestionConstSumSubmissionFormTemplate.html]: My bad, it's `width-auto` not `auto-width` :P

**Comment by wkurniawan07** [comment]
@joanneong in case you haven't realized yet (after 6+ days? 🤔), there's a merge conflict here to be resolved.

**Comment by joanneong** [comment]
Here is what we have after all the changes since I first opened my pull request:

**Distribute points among recipients**

image

**Distribute points among options (self)**

image

**Distribute points among options (others)**

image

**Distribute points (with long options)**

image

From this I've realised that UI changes are pretty subjective, and that it is hard to satisfy all possible scenarios. For example, my original pull request had a longer span for the status message, which was slightly uglier but could accommodate lengthy status messages. Whereas now it spans only the question and answer, which looks somewhat nicer but causes longer status message to run onto the next line. In any case, thanks a lot for your patient guidance @xpdavid !

**Comment by xpdavid** [comment]
@joanneong You can do the same fix for the rank question now. Create an issue and I'll help you update it :)

**Comment by joanneong** [comment]
@xpdavid Just to double check, so I should open a new issue for rank options questions? This is because in the discussion in #7899 @damithc suggested fixing the rank options questions in the same pull request (i.e. this pull request)...

P/S Sorry if this was mentioned before, but what does the `e.1` label mean? I can't seem to find it on the [issues guide](https://github.com/TEAMMATES/teammates/blob/master/docs/issues.md)...

**Comment by xpdavid** [comment]
@joanneong It's OK to open another PR (separation of concerns right? :P)

Don't worry the `e.*` now.
FYI: https://github.com/TEAMMATES/teammates/blob/master/docs/process.md#merging-a-pr

**Pull request merged**